workspace/resource_repo.go

```diff
@@ -0,0 +1,216 @@
+			Required: true,
+			ForceNew: true,
+		},
+		"git_provider": {
+			Type:     schema.TypeString,
+			Optional: true,
+			Computed: true,
+			ForceNew: true,
+			DiffSuppressFunc: func(k, old, new string, d *schema.ResourceData) bool {
+				return strings.EqualFold(old, new)
+			},
+		},
+		"path": {
+			Type:     schema.TypeString,
+			Optional: true,
+			Computed: true,
+			ForceNew: true, // TODO: remove after the Update API will support changing the path
+		},
+		"branch": {
+			Type:          schema.TypeString,
+			Optional:      true,
+			Computed:      true,
+			ConflictsWith: []string{"tag"},
+		},
+		"tag": {
+			Type:          schema.TypeString,
+			Optional:      true,
+			ConflictsWith: []string{"branch"},
+		},
+		"commit_hash": {
+			Type:     schema.TypeString,
+			Computed: true,
+		},
+	}
+	return common.Resource{
+		Schema:        s,
+		SchemaVersion: 1,
+		Create: func(ctx context.Context, d *schema.ResourceData, c *common.DatabricksClient) error {
+			reposAPI := NewReposAPI(ctx, c)
+			url := d.Get("url").(string)
+			provider := d.Get("git_provider").(string)
+			if provider == "" { // trying to infer Git Provider from the URL
+				provider = getProviderFromUrl(url)
+			}
+			if provider == "" {
+				return fmt.Errorf("git_provider isn't specified and we can't detect provider from URL")
+			}
+			req := createRequest{Path: d.Get("path").(string), Provider: provider, Url: url}
+			resp, err := reposAPI.Create(req)
+			if err != nil {
+				return err
+			}
+			d.SetId(fmt.Sprintf("%d", resp.Id))
+			branch := d.Get("branch").(string)
+			tag := d.Get("tag").(string)
+			if tag != "" {
+				d.Set("branch", "")
+				return reposAPI.Update(d.Id(), map[string]string{"tag": tag})
+			} else if branch != "" && branch != resp.Branch {
+				return reposAPI.Update(d.Id(), map[string]string{"branch": branch})
+			}
+
+			return nil
+		},
+		Read: func(ctx context.Context, d *schema.ResourceData, c *common.DatabricksClient) error {
+			reposAPI := NewReposAPI(ctx, c)
+			resp, err := reposAPI.Read(d.Id())
+			if err != nil {
+				return err
+			}
+			d.SetId(fmt.Sprintf("%d", resp.Id))
+			d.Set("url", resp.Url)
+			d.Set("path", resp.Path)
+			d.Set("git_provider", resp.Provider)
+			d.Set("commit_hash", resp.HeadCommitId)
+			if d.Get("branch").(string) == "" {
+				d.Set("branch", resp.Branch)
+			}
+			return nil
+		},
+		Update: func(ctx context.Context, d *schema.ResourceData, c *common.DatabricksClient) error {
+			// TODO: update may change ONE OF (url AND provider (optional)), (path), or (branch OR tag).
+			// for URL/provider force re-create as there are limits on what could be done for changing URL/provider
+			reposAPI := NewReposAPI(ctx, c)
+			// Not working yet, wait until API is ready
+			// if d.HasChange("path") {
+			// 	err := reposAPI.Update(d.Id(), map[string]string{"path": d.Get("path").(string)})
+			// 	if err != nil {
+			// 		return err
+			// 	}
+			// }
+			if d.HasChange("branch") || d.HasChange("tag") {
```

Comment:
If all non-branch/tag fields are forcenew, then this method is way shorter

Reply:
it's temporary. there are plans to extend API to allow to change other parameters without recreation

Reply:
then ForceNews will be removed?..

Reply:
yes, it will be - I put the comment for that field to remove when it's available.